File: BFS/src/flightpath320/BfsShortestPath.java
package flightpath320;

import java.util.Hashtable;
import java.util.LinkedList;
import java.util.Queue;
import java.security.KeyException;
import java.util.ArrayList;
import java.util.HashSet;

import java.util.UUID;

class ShortestPathBFS {
  static ArrayList<Node> run(Graph g, Node s, Node e) {
    Queue<Node> queue = new LinkedList<>();
    queue.add(s);

    s.isVisited = true;
    while (!queue.isEmpty()) {
      Node v = queue.remove();
      for (Node u : g.get(v)) {
        if (!u.isVisited) {
          u.setPred(v);
          u.isVisited = true;
          queue.add(u);
        }
        if (u.equals(e)) {
          return u.traceback(s);
        }
      }
    }

    return null;
  }
}

class Node {
  public String name;
  private Node pred;
  public boolean isVisited;

  public Node(String name) {
    this.pred = null;
    this.name = name;
    if (this.name == "dummy") {
      this.name = "dummy" + " " + UUID.randomUUID().toString();
    }
  }

  public Node() {
    this("dummy");
  }

  public Node getPred() {
    return pred;
  }

  public void setPred(Node node) {
    pred = node;
  }

  public ArrayList<Node> traceback(Node goal) {
    assert (pred != null);

    Node pred = this.pred;
    ArrayList<Node> path = new ArrayList<Node>();
    path.add(this);

    try {
      while (pred != goal) {
        if (goal != null && pred == null) {
          String msg = "node " + this + " has no predecessor ";
          msg += " and is not the starting node";
          throw new Exception(msg);
        }

        path.add(pred);
        pred = pred.pred;
      }
      path.add(goal);
    } catch (Exception e) {
      e.printStackTrace();
    }

    return path;
  }

  @Override
  public int hashCode() {
    return name.hashCode();
  }

  @Override
  public boolean equals(Object node) {

    return name.equals(((Node) node).name);
  }

  @Override
  public String toString() {
    String str = "<Node ";
    str += name;
    str += ">";
    return str;
  }

}

class Graph extends Hashtable<Node, HashSet<Node>> {

  private static final long serialVersionUID = 1L;

  Graph(ArrayList<Node> nodes) {
    super();

    for (Node n : nodes) {
      put(n, new HashSet<>());
    }

  }

  void addEdge(Node node1, Node node2) {
    HashSet<Node> node1Adjs = get(node1);
    if (!node1Adjs.contains(node2)) {
      node1Adjs.add(node2);
    }
  }

  void addEdge(Node node1, Node node2, int weight) throws Exception {
    HashSet<Node> node1Adjs = get(node1);
    if (node1Adjs == null) {
      throw new KeyException("node not in graph: " + node1);
    }

    if (!containsKey(node2)) {
      throw new KeyException("node not in graph: " + node2);
    }

    Node dummy = node1;
    while (weight != 0) {
      dummy = addDummy(dummy);
      weight -= 1;
    }
    assert (weight == 0);
    addEdge(dummy, node2);
  }

  Node addDummy(Node node) {
    Node dummy = new Node();
    put(dummy, new HashSet<>());
    addEdge(node, dummy);
    return dummy;
  }
}

File: main/bfs-shortest-path.py
# using bfs on unweighted graph to simulate weighted graph and
# to find the shortest path.


import collections as c
import typing as t
import uuid


class Node:
    def __init__(self, name="dummy"):
        self.pred: t.Optional['Node'] = None
        self.name: str = name
        self.is_airport: bool = False

        if name == "dummy":
            # dummies have unique nodes
            self.name = self.name + " " + str(uuid.uuid4())

    def traceback(self, goal: 'Node' = None) -> t.List['Node']:
        path = [self]
        pred = self.pred

        def stopping(v):
            if goal is not None and pred is None:
                raise RuntimeError(
                    f"node {self} has no predecessor " +
                    " and is not the starting node")

            return v is None if goal is None else v == goal

        while not stopping(pred):
            path.append(pred)
            pred = pred.pred
        path.append(goal)
        return path

    def __hash__(self):
        """
        Node will be unique by their name on a set
        """
        return hash(self.name)

    def __str__(self):
        return f"<Node {self.name}>"

    def __eq__(self, other):
        """
        Although if two nodes have the same name we consider them
        equal, only the one in the graph matters.
        """
        return self.name == other.name

    __repr__ = __str__


class Graph(c.UserDict):
    def __init__(self, nodes: t.List):
        super().__init__()

        for n in nodes:
            self.__setitem__(Node(n), set())

    def add_edge(self, node1: Node, node2: Node, weight=0):
        """
        Never add nodes here. Nodes need to exists before adding edges.
        """

        # node1_adjacents: t.Set = self.get(node1)
        node1_adjacents: t.Set = self[node1]
        if node1_adjacents is None:
            raise KeyError(f"node not in graph. Node: {node1}")
        if node2 not in self:
            raise KeyError(f"node not in graph. Node: {node2}")
        if weight == 0:
            if node2 not in node1_adjacents:
                node1_adjacents.add(node2)
        else:
            dummy = node1
            while weight != 0:
                dummy = self.add_dummy(dummy)
                weight -= 1
            assert weight == 0
            self.add_edge(dummy, node2)

    def add_dummy(self, node: Node) -> Node:
        """
        Add dummy nodes to simulate the distance.
        """
        dummy = Node()
        self[dummy] = set()
        self.add_edge(node, dummy)
        return dummy


def bfs(g: Graph, s: Node, e: Node):
    queue = [s]
    visited = [s]

    while queue != []:
        v = queue.pop(0)
        for u in g[v]:
            if u not in visited:
                u.pred = v
                visited.append(u)
                queue.append(u)
            if u == e:
                return u.traceback(s)
    return []


if __name__ == "__main__":
    g = Graph([1, 2, 3, 4])

    g.add_edge(Node(1), Node(3), weight=10)
    g.add_edge(Node(2), Node(3), weight=3)
    g.add_edge(Node(3), Node(4), weight=3)
    g.add_edge(Node(1), Node(4), weight=30)

    print("graph: ")
    __import__('pprint').pprint(g)

    visited = bfs(g, Node(1), Node(4))
    print("shortest path: ")
    __import__('pprint').pprint(visited)
